### ISSUE TITLE
Symbolic loops (like "scan" in Theano)

### ISSUE THREAD

**Issue by sisp** [tensorflow/tensorflow#208]
I've been wondering if there are plans to add symbolic loops to TensorFlow because I feel like this is a major feature when it comes to variable length sequences. Finite unfolding (with bucketing) seems like a dirty hack to me and since (as far as I understand) TensorFlow is meant for deployment, too, how do you envision using it for seq2seq translation given that you don't know ahead of time how long the generated sequence will be?

Thanks,
Sigurd

**Comment by vrv** [comment]
Our white paper mentions a number of control flow operations that we've experimented with -- I think once we're happy  with its API and confident in its implementation we will try to make it available through the public API -- we're just not quite there yet.  It's still early days for us :)

**Comment by sisp** [comment]
I see. Great news that that's in the pipeline. Can't wait until it's publicly available. ;-)

**Comment by dimchansky** [comment]
+1

**Comment by ebrevdo** [comment]
See the highly alpha and unsupported control_flow_ops.While, the
TensorArray python class, and nn.dynamic_rnn.  All at HEAD.

**Comment by myme5261314** [comment]
Any update or progress schedule?

**Comment by ebrevdo** [comment]
You can now use control_flow_ops.{map, foldl, foldr}  with forward and backprop, and you can call these functions from inside each others' lambdas.  For RNN, you can use dynamic_rnn which does the same.  If you have more complex eneds you can comment them here or use control_flow_ops.While and TensorArray.  I'm marking this as fixed.

**Comment by vrv** [comment]
@ebrevdo: are those functions:

1) public
2) documented?

**Comment by ebrevdo** [comment]
Indeed not, but I believe {map_fn, foldl, foldr} are ready to be added by
referencing them in the header.  We'll get a CL out soon.

**Comment by rakeshvar** [comment]
A toy RNN example with these functions would go a long way; especially since they do not directly correspond to the `theano.scan` functionality.

**Comment by ebrevdo** [comment]
See the implementation of dynamic_rnn for a comprehensive example.

**Comment by petrbel** [comment]
Apart from `map` and `fold`s, do you plan to implement also `scan` (as the original question mentions)?

**Comment by vrv** [comment]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/functional_ops.py#L256 !

**Comment by rdipietro** [comment]
@rakeshvar here is a vanilla-RNN example: https://nbviewer.jupyter.org/github/rdipietro/tensorflow-notebooks/blob/master/tensorflow_scan_examples/tensorflow_scan_examples.ipynb

(`dynamic_rnn` currently uses `While` and `TensorArray`, not `scan`.)

**Comment by ebrevdo** [comment]
Scan uses the same approach as dynamic_rnn.  The latter was made first.
Whereas in theano, scan was the fundamental loop method, in tensorflow that
behavior is subsumed my while loop and tensorarray.
On Apr 7, 2016 2:45 PM, "Robert DiPietro" <notifications@github.com> wrote:

@rakeshvar <https://github.com/rakeshvar> here is a vanilla-RNN example:
https://nbviewer.jupyter.org/github/rdipietro/tensorflow-notebooks/blob/master/tensorflow_scan_examples/tensorflow_scan_examples.ipynb

(dynamic_rnn currently uses While and TensorArray, not scan.)

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/tensorflow/tensorflow/issues/208#issuecomment-207100390>

**Comment by 22csnyder** [comment]
@rdipietro  The link to that notebook does not work for me :(

**Comment by rdipietro** [comment]
Oh I'm sorry. Reorganized a bit.

http://rdipietro.github.io/tensorflow-scan-examples/

**Comment by rajarsheem** [comment]
@rdipietro I tried to extend your code to a vanilla GRU. Also removed the clipping therefore and replaced Gradient Descent with Adagrad.
Consequently convergence is now much faster.
Can you please verify the code ?
http://paste.ubuntu.com/16534925/
(I know it's much dirty!)

**Comment by KnHuq** [comment]
How to use map and scan together??????? Anyone have any idea?

x=tf.constant(
[[[1,2,3],[10,20,30]],
[[1,2,3],[10,20,30]],
[[1,2,3],[10,20,30]], 
[[1,2,3],[10,20,30]]])

def sum(x): return  tf.scan(lambda y, z : tf.add(y,z), x)
tf.map_fn(sum ,a )

Then errors pop up!!!! Is there any way using these together?
@rdipietro I have been through your example notebook but how to use it for batch????

**Comment by yuanbyu** [comment]
The implementation of both scan and map_fn uses TensorArray. Nesting scan inside map_fn is essentially nested while loops. Unfortunately, there is a known bug when TensorArray and nested while loops are used together.  We have been working on a fix.

**Comment by KnHuq** [comment]
Ok thanks.

**Comment by rajarsheem** [comment]
@KnHuq Not sure if this serves your problem for batches, but you can still see my [attempt](https://github.com/rajarsheem/playing-with-rnns/blob/master/rnn.ipynb) . There may be some issues, though.

**Comment by KnHuq** [comment]
I am using scan to deal with different sequence length. But having problem to deal with batch! @rajarsheem can you write your solution for my above  problem??? If I use for loop then its same like online learning! No need to use batch then! So no need to use GPU then!

Please Google solve this nested while loops. ...

**Comment by rdipietro** [comment]
You can see [here](https://github.com/rdipietro/miccai-2016-surgical-activity-rec/blob/master/train_and_eval.ipynb) for a more complex example, with a batch size > 1. But you'll need to either a) write not-so-clean code that's efficient (as they do in TensorFlow officially) or b) write clean code that is less efficient (this is what I do; I just wrap shorter sequences in time until all sequences are the same length, which simultaneously makes short sequences "count" just as much loss wise as long sequences).

**Comment by KnHuq** [comment]
@rdipietro Thanks a lot. I will go through the your example!

**Comment by rajarsheem** [comment]
@rdipietro can you verify my graph, whether it is set up right ? (it is a very very simple [example](https://github.com/rajarsheem/playing-with-rnns/blob/master/rnn.ipynb))

**Comment by KnHuq** [comment]
Finally I have figured out how to use scan with batch. So I have written some simple code [RNN](http://knhuq.github.io/RNN.html) and [GRU](http://knhuq.github.io/GRU.html). Have conducted a little hack as Tensorflow was raising low level error.  The github repo is [here](https://github.com/KnHuq/Dynamic_RNN_Tensorflow)

**Comment by KnHuq** [comment]
The problem of scan and map together is solved by the latest release of Tensorflow.
